File: frontend/ClickAndDrive/src/app/services/user.service.ts
// This service will be used for all user interactions
// Get a user, change his informations, change password...

import { HttpClient } from "@angular/common/http";
import { Inject, Injectable } from "@angular/core";
import { UserProfileInformation } from "./models/user-profile-information";
import { Observable } from "rxjs";
import { VehicleInformation } from "./models/driver-vehicle-information";
import { ChangePasswordRequest } from "./models/change-password";
import { ProfileImageResponse } from "./models/profile-image-response";

@Injectable({providedIn: 'root'})
export class UserService {

    constructor(private http: HttpClient) {}

    private apiUrlUser = 'http://localhost:8080/api/user';
    private apiUrlDriver = 'http://localhost:8080/api/drivers';

    // Get logged in users information for display and change
    getUserProfileInfo(userId: number): Observable<UserProfileInformation> {
        return this.http.get<UserProfileInformation>(`${this.apiUrlUser}/${userId}/profile`);
    }

    // Get vehicle information for driver
    getDriverVehicle(userId: number): Observable<VehicleInformation> {
        return this.http.get<VehicleInformation>(`${this.apiUrlDriver}/${userId}/vehicle`);
    }

    // Save changed information (if driver + vehicle)
    changeUserInfo(userId: number, updatedProfile: UserProfileInformation): Observable<UserProfileInformation> {
        return this.http.put<UserProfileInformation>(`${this.apiUrlUser}/${userId}/profile-update`, updatedProfile);
    }

    // Upload a profile photo
    uploadProfileImage(userId: number, formData: FormData): Observable<ProfileImageResponse> {
        return this.http.post<ProfileImageResponse>(`${this.apiUrlUser}/${userId}/profile-image`, formData);
    }

    // Delete users profile photo
    deleteProfileImage(userId: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrlUser}/${userId}/delete-profile-image`);
    }

    // Change user password
    changePassword(request: ChangePasswordRequest): Observable<any> {
        return this.http.post(`${this.apiUrlUser}/change-password`, request);
    }
}

File: mobile-version/Navbar-shortened/app/src/main/java/com/example/clickanddrive/clients/services/UserService.java
package com.example.clickanddrive.clients.services;

import com.example.clickanddrive.dtosample.requests.ChangePasswordRequest;
import com.example.clickanddrive.dtosample.requests.UpdateProfileRequest;
import com.example.clickanddrive.dtosample.responses.ProfileImageResponse;
import com.example.clickanddrive.dtosample.responses.UserProfileResponse;

import java.util.Map;

import okhttp3.MultipartBody;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.Path;

public interface UserService {

    // Get user profile information
    @GET("user/{id}/profile")
    Call<UserProfileResponse> getUserProfile(@Path("id") Long id);

    // Update user profile information
    @PUT("user/{id}/profile-update")
    Call<UserProfileResponse> updateUserProfile(@Path("id") Long id, @Body UpdateProfileRequest request);

    // Change password
    @POST("user/change-password")
    Call<Map<String, String>> changePassword(@Body ChangePasswordRequest request);

    // Allow user to upload profile photo
    @Multipart
    @POST("user/{id}/profile-image")
    Call<ProfileImageResponse> uploadProfileImage(@Path("id") Long userId, @Part MultipartBody.Part file);

    // Allow user to remove profile photo
    @DELETE("user/{id}/delete-profile-image")
    Call<Void> deleteProfileImage(@Path("id") Long userId);

    // Get profile photo
    @GET("user/profile-images/{filename}")
    Call<ResponseBody> getProfileImage(@Path("filename") String filename);

}
